src/Dev/Validation/RelationValidationService.php

```diff
@@ -0,0 +1,639 @@
+     *
+     * @var array
+     */
+    private static $allow_rules = [
+        'empty' => '',
+        'app' => 'App',
+    ];
+
+    /**
+     * Any classes with the following namespaces/class prefixes will not be inspected
+     * This config is intended to be used together with @see $allow_rules to narrow down the inspected classes
+     * Empty string is a special value which represents classes without namespaces
+     * Set value to null to disable the rule (useful when overriding configuration)
+     *
+     * @var array
+     */
+    private static $deny_rules = [];
+
+    /**
+     * Relations listed here will not be inspected
+     * Format is <class>.<relation>
+     * for example: Page::class.'.LinkTracking'
+     *
+     * @var array
+     */
+    private static $deny_relations = [];
+
+    /**
+     * Ignore any configuration, useful for debugging specific classes
+     *
+     * @var bool
+     */
+    protected $ignoreConfig = false;
+
+    /**
+     * @var array
+     */
+    protected $errors = [];
+
+    public function flushErrors(): void
+    {
+        $this->errors = [];
+    }
+
+    public static function reset(): void
+    {
+        $service = self::singleton();
+        $service->flushErrors();
+        $service->ignoreConfig = false;
+    }
+
+    /**
+     * @throws ReflectionException
+     */
+    public static function flush(): void
+    {
+        self::singleton()->executeValidation();
+    }
+
+    /**
+     * Hook this into your unit tests and assert for empty array like this
+     *
+     * $messages = RelationValidationService::singleton()->validateRelations();
+     * $this->assertEmpty($messages, print_r($messages, true));
+     *
+     * @return array
+     * @throws ReflectionException
+     */
+    public function validateRelations(): array
+    {
+        self::reset();
+        $classes = ClassInfo::subclassesFor(DataObject::class);
+
+        return $this->validateClasses($classes);
+    }
+
+    /**
+     * @throws ReflectionException
+     */
+    public function executeValidation(): void
+    {
+        $errors = $this->validateRelations();
+        $count = count($errors);
+
+        if ($count === 0) {
+            return;
+        }
+
+        DB::alteration_message(
+            sprintf(
+                '%s : %d issues found (listed below)',
+                ClassInfo::shortName(static::class),
+                $count
+            ),
+            'notice'
+        );
+
+        foreach ($errors as $message) {
+            DB::alteration_message($message, 'notice');
+        }
+    }
+
+    /**
+     * Inspect specified classes - this ignores any configuration
+     * Useful for checking specific classes when trying to fix relation configuration
+     *
+     * @param array $classes
+     * @return array
+     */
+    public function inspectClasses(array $classes): array
+    {
+        self::reset();
+        $this->ignoreConfig = true;
```

Comment:
should `$this->ignoreConfig = false; be called after calling `$this->validateClasses($classes);` ?

Reply:
It needs to be called before the validation as configuration potentially narrows down the scope of the covered classes. This is meant to be used for debugging / figuring out configuration.

Reply:
Reset to previous state of this flag is covered by `self::reset()` so there is no need to set it back.